Study animal df templates

fkang-pu opened the pull request:
<!-- markdownlint-disable-next-line first-line-heading -->
## Summary Change Description

- Created a list of Pandas DataFrames in utils.py

- Added a customtag for object hyperlinks (studies, tracers, and treatments)

- Added ListView and DetailView for Tissue

- Added a function-based view and template (study_summary.html) for study-based summary table
 (including all studies, animals, samples)

- Modified the following views and the templates"
 StudyListView, StudyDetailView,
 AnimalListView, AnimalDetailView
 SampleListView, SampleDetailView
 using Pandas DataFrames to get list and stats data for studies, animal, samples


## Affected Issue Numbers

- Resolves #211 
- Resolves #212
- Resolves #214
- Resolves #215
- Resolves #216

## Code Review Notes

- Formatted values for sample_date, mrun_date, sample_collect_time in DataFrames, in order to display in correct
 format after converting data using pd.to_json.

- Some animals have no treatment data, and thus no link to protocols.
 Pandas DataFrame converted int field for treatment_id to float (due to null values). To work around it, assigned treatment_id to 0 for rows if treatment is null in DataFrame.

- need to add/modify test code for Pandas DataFrames and views. But like to get 1st round of review to see if the strategy for using DataFrames and the code in utils.py and views.py are OK.
 
- tried to construct Querysets and DataFrames for efficient data filtering and join. But need more tests and tuning with larger datasets.

## Checklist

- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- [ ] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
- [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)

lparsons reviewed (commented):
Overall this looks great, but there are a few things that I feel would be worth a discussion of the implementation: 

* The conversion of records to data frames are `classmethods` in `utils`. I would suggest that they either be moved into their own module or into the models themselves (perhaps also splitting the models up into modules would be helpful at this point). The reason is that this would keep the tightly coupled things (lists of specific field names) together.
* Using a `get_absolute_url()` function on the models would greatly simplify the filter that generates lists of objects as links. It would also have the benefit of being utilized by the admin interface.
* Tests for the new views. In some cases, I suspect there are already tests (lists of samples, etc.) but if there aren't (new Tissue page) ones should be added. It would also be nice to possibly test a few of the summary stats in the existing view stats.

I don't necessarily think these all have to be addressed in this PR, and I'm not entirely sure I have a complete understanding of the ramifications of my suggestions, but I think a discussion about these issues would be helpful.

fkang-pu commented:
@lparsons I will keep your comments and go through them again when making modifications. Just answer your questions in general:
The questions you asked I also debated with myself several times. We should have further discussions about the following issues, though they are not top priorities for beta release. 

- It's fine to move data frames to a separate module, but it's not my preferred choice to move them into 
Models since some of them are shared among Models (e.g. all_anim_msrun_df contains combined data from Study, 
Animal, Tissue, Sample, and MSRun, called by several views). I also feel it's better to break models.py, views.py, utils.py into several files in sub-folders in the future.

- I like to have get_absolute_url() for every model. I raised this question a few months ago, but seems we decided not to do it, use {% url 'object_detail' pk %} instead. We can discuss it again.

- About field names in DataFrames:  to use values_list() (or values() function ) in a queryset to get related data from other models, the fields names are formed based on query path, some of which have long names with multiple double underscores. It's easy to rename columns with Pandas, I tried to make them short, easy to read, and consistent for columns in different DataFrames. Basically, I  attempted to do data manipulation with Pandas DataFrames, and then use them with minimum work when rendering data in templates.

- my to-do list:

1) remove debug code
2) add test for tissue list and detail. 
3) modify test code for other views (study, animal, sample)  
Hopefully I'll find time to do it Friday afternoon (or early next week).

jcmatese commented on DataRepo/templates/DataRepo/sample_list.html:
```diff
@@ -1,57 +1,93 @@
+<br>
+    {% if df %}
+    <table class="table table-sm table-hover table-bordered table-responsive-xl table-striped"
+        id="table"
```
C: There are several `id="table"` definitions in the templates that probably won't cause problems (if there are only one identified "table" on the page), but _may_ cause an issue if another `id="table"` pops up on the same page in the future.  You _might_ want to be more specific in the identification, though I don't think it is a blocking issue.

jcmatese commented on DataRepo/utils.py:
```diff
@@ -1003,3 +1004,427 @@ class MissingSamplesError(Exception):
+        data = json.loads(json_records)
+        return data
+
+    @classmethod
```
Probably just the convenience of calling `cls.qs_to_df` .  It does _seem_ that _maybe_ some aspects of these methods could become the foundation of some abstract base class `PandasModel` , or similar,  where if a Django model also claimed inheritance, and also defined a attribute-header mapping, could inherit putative methods like `queryset_to_dataframe` or `dataframe_to_list_of_dictionaries` or `get_all_objects_dataframe`.  Would need more thought, though, because a QuerySet can be a pretty fluid thing, and not necessarily linked to a specific model.

lparsons commented:
@fkang-pu Thanks for the detailed responses. I think you have a good plan to get this PR wrapped up. We can discuss the potentially larger changes in upcoming developers meetings.

* I would agree that the data frame functions wouldn't necessarily fit with a specific model class, but that they do deserve their own module. I'm not convinced that a class full of static and/or class methods is the best way to do this, but we can discuss code reorganization after we get the beta release done. I'll add some of these notes to our developers meeting agendas going forward.
* I'm not sure why the `get_absolute_url()` methods were skipped, but I do think they have the potential to make that filter much cleaner. I'll add this to our agenda, and if there is general agreement, we'll create an issue for it.
* I can see how the field names need to get renamed, and it's not clear to me what the options are. Making sure that it is done independent of the template seems like a good idea. Perhaps we will learn additional options for this sort of thing going forward, but I like the idea of having a clear interface to the template, the template expects a dataframe with consistently named columns.  

Once you complete the items on your to-do list, send me a request for a re-review.

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -1003,3 +1004,427 @@ class MissingSamplesError(Exception):
+        per_anim_msrun_df = all_anim_msrun_df[
+            all_anim_msrun_df["animal_id"] == animal_id
+        ]
+        return per_anim_msrun_df
```
C. This isn't an issue.  I'm just commenting - because I don't really know. I haven't played with pandas dataframes much yet, but I'm unsure whether there's a more succinct way to accomplish what all these `_df` methods do, and if not, I think they might be a candidate for a separate file all together. I don't know for sure, but a lot of the things in these df methods feel like view decisions, and the methods feel more specialized than what I might expect in a utils method.  I.e. they produce output specific to one particular view and there *could* exist variants for other views.  I just thought I would mention these thoughts to see if anyone else has an opinion on it in general...

Also, there are some instances also where all the fields are passed explicitly to `values_list()`, in which case, you don't need to specify the field names (default behavior is all).

fkang-pu replied:
leave the code without change. Likely re-write some code when I have a better knowledge about Pandas functions.

hepcat72 commented on DataRepo/views.py:
```diff
@@ -28,9 +28,12 @@
 )
 from DataRepo.multiforms import MultiFormsView
-from DataRepo.utils import MissingSamplesError, ResearcherError
+from DataRepo.utils import MissingSamplesError
+from DataRepo.utils import QuerysetToPandasDataFrame as qs2df
+from DataRepo.utils import ResearcherError
```
NB. I bet there's a way to put these all in one call to import. At the least, lines 34 and 36 could be on one line.  @lparsons?

fkang-pu replied:
formatted this way as suggested by Superlinter.

hepcat72 replied:
I suspect that superlinter wouldn't have a problem with MissingSamplesError and ResearcherError on the same line, but I don't know for sure.  Resolving.

hepcat72 commented on DataRepo/utils.py:
```diff
@@ -1003,3 +1004,427 @@ class MissingSamplesError(Exception):
+
+        # some animal records have no treatment data, DataFrame converts the dtype from int to float
+        # workaround: set treatment_id to 0 for NaN, and convert dtype back to int
+        anim_list_df["treatment_id"] = (
+            anim_list_df["treatment_id"].fillna(0).astype(int)
+        )
```
Q. Can this be filled with `None`?  That would be consistent with other view renderings for fields not having a value.

fkang-pu replied:
need to format this column as datatype int order to make hyperlink work. None would cause Pandas to convert it to float automatically.  The code is working, so keep it until I have better solution.

hepcat72 replied:
Q. Not sure I understand, so let me put it in my own words and you correct me if I'm wrong.  So if we leave it as `None` (or fill with `None`), that `None` value is changed to `0.0`? Does it change other populated values in that column to floats as well? So you're filling with integer zeros and then catching those zeroes and omitting the link (or displaying `None`?  Either way, I'm resolving.  Just wanted to know how it worked.

fkang-pu replied:
@Rob if I filled with None, it changed the data type for the column. e.g if the column value is 1, it's changed to 1.0 which would mess up with hyperlink using id.

hepcat72 reviewed (approved):
There's a lot of great work in here Fan.  I like all of the added functionality and view improvements.  I do have some suggestions and some questions, but they are all non-blocking.

hepcat72 reviewed (changes requested):
Changing my review result.  I switched to your branch and started testing it out.  I hit an exception on the Studies page: `'NoneType' object has no attribute 'split'`:

```
Internal Server Error: /DataRepo/studies/
Traceback (most recent call last):
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/core/handlers/base.py", line 204, in _get_response
    response = response.render()
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/response.py", line 105, in render
    self.content = self.rendered_content
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/response.py", line 83, in rendered_content
    return template.render(context, self._request)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 170, in render
    return self._render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 162, in _render
    return self.nodelist.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/defaulttags.py", line 312, in render
    return nodelist.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/defaulttags.py", line 211, in render
    nodelist.append(node.render_annotated(context))
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 988, in render
    output = self.filter_expression.resolve(context)
  File "/Users/rleach/local/miniconda3/lib/python3.6/site-packages/django/template/base.py", line 698, in resolve
    new_obj = func(obj, *arg_vals)
  File "/Users/rleach/GoogleDrivePU/PROJECT/TRACEBASE/REPO/TraceBase/DataRepo/templatetags/customtags.py", line 141, in obj_hyperlink
    id_name_dict = dict(x.split("||") for x in id_name_list)
  File "/Users/rleach/GoogleDrivePU/PROJECT/TRACEBASE/REPO/TraceBase/DataRepo/templatetags/customtags.py", line 141, in <genexpr>
    id_name_dict = dict(x.split("||") for x in id_name_list)
AttributeError: 'NoneType' object has no attribute 'split'
```

I'll do some more testing and report anything else I encounter...

fkang-pu commented:
@hepcat72 No rush. Feel free to make more suggestions, as I won't have time for this project until Friday afternoon.  I didn't get the "NoneType" error you mentioned with studies.

hepcat72 commented:
> @hepcat72 No rush. Feel free to make more suggestions, as I won't have time for this project until Friday afternoon. I didn't get the "NoneType" error you mentioned with studies.

Yeah, it's data-related.  When I remove the obj_hyperlink filters on that page, my third row shows: `['1||obob_fasted', None]` in the treatment column.  So instead of:

```
    if id_name_list == [None]:
        return None
    else:
        id_name_dict = dict(x.split("||") for x in id_name_list)
        obj_format_html = format_html_join(
            ",",
            '<a href="{}">{}</a>',
            [
                (reverse(tmplt_name, args=[str(id)]), id_name_dict[id])
                for id in id_name_dict
            ],
        )
        return obj_format_html
```
You should probably have the conditional in the loop, something like (though I haven't tested this):
```
    if id_name_list == [None]:
        return None
    else:
        id_name_dict = {}
        for x in id_name_list:
            if x != None:
               k, v = x.split("||")
               id_name_dict[k] = v
        obj_format_html = format_html_join(
            ",",
            '<a href="{}">{}</a>',
            [
                (reverse(tmplt_name, args=[str(id)]), id_name_dict[id])
                for id in id_name_dict
            ],
        )
        return obj_format_html
```

fkang-pu commented:
@lparsons @hepcat72 Pushed changes after rebase. Please review again. I'll take another look after getting in office.

changes:

- removed debug code
- modified customtags.py
- added test_utils.py for verifying data in selected DataFrames.
- modified test_views.py for code related to Study, Tissue, Animal, and Sample
- minor changes to templates

notes:
Keep QuerysetToPandasDataFrame code in utils.py for now. I think it's better to modify/rewrite later when re-organizing file structures for DataRepo app.  May need to add additional DataFrames in the near future (e.g. get tracers from compound list). 

For question about staticmethod vs classmethod, I am still thinking about it. Also consider where is best place to put the code in order to cache some of the DadaFrames. Right now they can be cached in views if needed.  Need more time to have better understanding of Pandas functions and think about how to modify and re-organize the code. Feel free to take over the task if you have better ideas.

lparsons reviewed (approved):
This looks good and seems to be working to me. The tests appear to cover the basics, thanks for adding those. A few minor things:

* @hepcat72's [suggestion about adding units in some of the headers](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/241/files#r746773872)
* @hepcat72's issue with a [`NoneType` in the split](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/241#issuecomment-965726101) - I'm not clear on what data exposed this issue, but creating a test for that edge case might be worth it. If simple, it could be done in this PR, or perhaps a separate issue created for it.

Just to note, I'll go ahead and create some issues to track some other things to be tackled later, namely:
* Refactor the `QuerysetToPandasDataFrame` code
* Create `get_absoute_url()` functions and use them to simplify the `get_ojbhyperlink()` filter

lparsons commented:
> @lparsons @hepcat72 Pushed changes after rebase. Please review again. I'll take another look after getting in office.

Just and FYI, it can be helpful if you use the "request re-review" function in GitHub. That makes it somewhat easier to see what PRs are pending reviews when we login.

![image](https://user-images.githubusercontent.com/645128/142030206-93d5f7c7-6479-430b-8b95-f5d6ed8a5eb5.png)

fkang-pu commented on DataRepo/templates/DataRepo/animal_detail.html:
```diff
@@ -1,69 +1,155 @@
+                    </a>
+                </td>
+                <td>
+                    {% if animal.treatment %}
+                        <a href="{% url 'protocol_detail' animal.treatment_id %}">{{ animal.treatment }}</a>
+                    {% else %}
+                        {{ animal.treatment }}
+                    {% endif %}
```
It worked, though use "None" seems better. Made change as suggested.

fkang-pu commented on DataRepo/templates/DataRepo/animal_detail.html:
```diff
@@ -1,69 +1,155 @@
+                </td>
+                <td>{{ animal.tracer_labeled_atom }}</td>
+                <td>{{ animal.tracer_labeled_count }}</td>
+                <td>{{ animal.tracer_infusion_rate }}</td>
+                <td>{{ animal.tracer_infusion_concentration }}</td>
+                <td>{{ animal.body_weight }}</td>
```
I would prefer all columns values left aligned. small issue, deal with later if you all want it right-align them.

hepcat72 replied:
Why is that your preference? Right-aligning number-columns is a convention that allows the places match up.  At a glance, you can compare numbers without having to count digits.  The only reason I can think of left-aligning would be if all values are less than 1 and a variable number of decimal places were used.  Besides, advanced search already right-aligns and I think we should be consistent.  We can change it after this PR.  This is non-blocking.  I'm just curious as to your preference...

fkang-pu replied:
I didn't think too much about this issue, but thought a table looks better if both column names and values all align left.

fkang-pu commented on DataRepo/templates/DataRepo/animal_detail.html:
```diff
@@ -1,69 +1,155 @@
+            {% for i in df %}
+            <tr>
+                <td><a href="{% url 'animal_detail' i.animal_id %}">{{ i.animal }}</a></td>
+                <td><a>{{ i.study_id_name_list |obj_hyperlink:"study" }}</a></td>
```
removed <a> tag, it was an error. still use customtag, easier read and handle in templates.

fkang-pu commented on DataRepo/templates/DataRepo/sample_detail.html:
```diff
@@ -25,12 +25,24 @@ <h4>Sample Record - {{ sample.name }}</h4>
         <tr>
             <td>Tissue</td>
-            <td>{{ sample.tissue}}</td>
+            <td><a href="{% url 'tissue_detail' sample.tissue.id %}">{{ sample.tissue }}</a></td>
```
fixed now

fkang-pu commented:
> > @hepcat72 No rush. Feel free to make more suggestions, as I won't have time for this project until Friday afternoon. I didn't get the "NoneType" error you mentioned with studies.
> 
> Yeah, it's data-related. When I remove the obj_hyperlink filters on that page, my third row shows: `['1||obob_fasted', None]` in the treatment column. So instead of:
> 
> ```
>     if id_name_list == [None]:
>         return None
>     else:
>         id_name_dict = dict(x.split("||") for x in id_name_list)
>         obj_format_html = format_html_join(
>             ",",
>             '<a href="{}">{}</a>',
>             [
>                 (reverse(tmplt_name, args=[str(id)]), id_name_dict[id])
>                 for id in id_name_dict
>             ],
>         )
>         return obj_format_html
> ```
> 
> You should probably have the conditional in the loop, something like (though I haven't tested this):
> 
> ```
>     id_name_dict = {}
>     for x in id_name_list:
>         if x != None:
>            k, v = x.split("||")
>            id_name_dict[k] = v
>     obj_format_html = format_html_join(
>         ",",
>         '<a href="{}">{}</a>',
>         [
>             (reverse(tmplt_name, args=[str(id)]), id_name_dict[id])
>             for id in id_name_dict
>         ],
>     )
>     return obj_format_html
> ```

I edited my code in customtags.py based on your suggestion. Hopefully it fixed "NoneType" error. I don't have a case in my local tracebase, but tested a similar case in Django shell.

hepcat72 reviewed (approved):
A few unaddressed issues, but all non-blocking.  Good job Fan.
